lib/Language/Souffle/Interpreter.hs

```diff
@@ -0,0 +1,237 @@
+data HandleData prog = HandleData
+  { soufflePath :: FilePath
+  , basePath    :: FilePath
+  , factPath    :: FilePath
+  , outputPath  :: FilePath
+  , datalogExec :: FilePath
+  , noOfThreads :: Word64
+  }
+
+newtype Handle prog = Handle (IORef (HandleData prog))
+
+instance MonadSouffle SouffleM where
+  type Handler SouffleM = Handle
+
+  {- | Looks for a souffle interpreter.
+
+     The action will return 'Nothing' if it failed to load the Souffle program.
+     Otherwise it will return a 'Handle' that can be used in other functions
+     in this module.
+  -}
+  init :: forall prog . Program prog => prog -> SouffleM (Maybe (Handle prog))
+  init prg = SouffleM $ do
+    -- TODO: Use system commands to locate the interpreter
+    -- TODO: Use system's tmp directory
+    tmpDir <- getCanonicalTemporaryDirectory
+    souffleTempDir <- createTempDirectory tmpDir "souffle"
+    let factDir = souffleTempDir </> "fact"
+    let outDir  = souffleTempDir </> "out"
+    createDirectoryIfMissing True $ factDir
+    createDirectoryIfMissing True $ outDir
+    datalogExecutable <- datalogProgramFile prg
+    souffleBin  <- souffleCmd
+    fmap (Just . Handle) $ newIORef $ HandleData
+      { soufflePath = souffleBin
+      , basePath    = souffleTempDir
+      , factPath    = factDir
+      , outputPath  = outDir
+      , datalogExec = datalogExecutable
+      , noOfThreads = 1
+      }
+
+  -- | Runs the Souffle program.
+  run :: Handle prog -> SouffleM ()
+  run (Handle ref) = SouffleM $ do
+    handle <- readIORef ref
+    -- Invoke the souffle binary using parameters, supposing that the facts are placed
+    -- in the factPath, rendering the output into the outputPath
+    callCommand $
+      printf "%s -F%s -D%s -j%d %s"
+        (soufflePath handle)
+        (factPath handle)
+        (outputPath handle)
+        (noOfThreads handle)
+        (datalogExec handle)
+
+  -- | Sets the number of CPU cores this Souffle program should use.
+  setNumThreads :: Handle prog -> Word64 -> SouffleM ()
+  setNumThreads (Handle ref) n = SouffleM $ liftIO $
+    modifyIORef' ref (\h -> h { noOfThreads = n })
+
+  -- | Gets the number of CPU cores this Souffle program should use.
+  getNumThreads :: Handle prog -> SouffleM Word64
+  getNumThreads (Handle ref) = SouffleM $ liftIO $
+    noOfThreads <$> readIORef ref
+
+  -- | Load all facts from files in a certain directory.
+  loadFiles :: Handle prog -> FilePath -> SouffleM ()
+  loadFiles (Handle ref) srcPath = SouffleM $ liftIO $do
+    handle  <- readIORef ref
+    let destPath = factPath handle
+    copyFiles srcPath destPath
+
+  -- | Write out all facts of the program to CSV files
+  --   (as defined in the Souffle program).
+  -- TODO: This should be moved out from this class
+  writeFiles :: Handle prog -> {- FilePath -> -} SouffleM ()
+  writeFiles (Handle ref) {-destPath-} = SouffleM $ liftIO $do
+    handle <- readIORef ref
+    let srcPath = factPath handle
+    copyFiles srcPath "./facts"
```

Comment:
Maybe `loadFiles` and `writeFiles` should have their own type classes. As it doesn't mean too much for the interpreted souffle. What do you think?

Reply:
Yeah, it's only useful for compiled version, good point.